inbm/dispatcher-agent/dispatcher/sota/oras_util.py

```diff
@@ -0,0 +1,187 @@
+        data = json.loads(response.text)
+        logger.debug(f"resp={data}")
+        # Calculate the total size
+        file_size = 0
+        for layer in data['layers']:
+            file_size += layer['size']
+        logger.debug(f"Total file size: {file_size}")
+
+    except (TypeError, KeyError, json.JSONDecodeError, SotaError) as err:
+        err_msg = f"Error getting artifact size from {manifest_uri} using token={jwt_token} Error: {err}"
+        logger.error(err_msg)
+        raise SotaError(err_msg)
+
+    if destination_repo.exists():
+        get_free_space = destination_repo.get_free_space()
+        free_space: int = int(get_free_space)
+    else:
+        raise SotaError("Repository does not exist : " +
+                                  destination_repo.get_repo_path())
+
+    logger.debug("get_free_space: " + repr(get_free_space))
+    logger.debug("Free space available on destination_repo is " + repr(free_space))
+    logger.debug("Free space needed on destination repo is " + repr(file_size))
+    return True if free_space > file_size else False
+
+def parse_uri(uri: CanonicalUri) -> Tuple[str, str, str, str, str, str, str]:
+    """ Parse the uri.
+
+    In case of uri.value = https://registry-rs.internal.ledgepark.intel.com/one-intel-edge/tiberos:latest
+    source = https://registry-rs.internal.ledgepark.intel.com/one-intel-edge
+    registry_server = registry-rs.internal.ledgepark.intel.com
+    image = tiberos
+    image_tag = latest
+    image_full_path = registry-rs.internal.ledgepark.intel.com/one-intel-edge/tiberos:latest
+    repository_name = one-intel-edge
+    registry_manifest = https://registry-rs.internal.ledgepark.intel.com/v2/one-intel-edge/tiberos/manifest/latest
+
+    @return: str representing source, registry_server, image, image_tag, image_full_path, repository_name and
+             registry_manifest
+    """
+    source = uri.value[:-(len(uri.value.split('/')[-1]) + 1)]
+    parsed_uri = urlparse(uri.value)
+    registry_server = parsed_uri.netloc
+    parsed_uri.geturl()
+    path_parts = parsed_uri.path.strip('/').split('/')
+    if len(path_parts) < 2:
+        raise SotaError(f"Invalid URI format: {uri.value}")
+    repository_name = '/'.join(path_parts[:-1])
+    image = path_parts[-1].split(':')[0]
+    image_tag = path_parts[-1].split(':')[1] if ':' in path_parts[-1] else 'latest'
+    image_full_path = f"{registry_server}/{repository_name}/{image}:{image_tag}"
+    registry_manifest = f"https://{registry_server}/v2/{repository_name}/{image}/manifests/{image_tag}"
+
+    return source, registry_server, image, image_tag, repository_name, image_full_path, registry_manifest
+
+def read_oras_token() -> str:
+    """Read oras JWT token from a path configured by Tiber OS node-agent. The node agent will renew the token when
+    the token is expired.
+
+    @return: JWT token to access release server
+    """
+    token = None
+    if os.path.exists(ORAS_TOKEN_PATH):
```

Comment:
handle possible OSError from these os commands/from read

Reply:
Added try/except to catch the OSError